providers/implementations/keymgmt/dh_kmgmt.c

```diff
@@ -128,70 +131,83 @@ static int key_to_params(DH *dh, OSSL_PARAM_BLD *tmpl)
     return 1;
 }
 
-static void *dh_importdomparams(void *provctx, const OSSL_PARAM params[])
+static void *dh_newdata(void *provctx)
 {
-    DH *dh;
-
-    if ((dh = DH_new()) == NULL
-        || !params_to_domparams(dh, params)) {
-        DH_free(dh);
-        dh = NULL;
-    }
-    return dh;
+    return DH_new();
 }
 
-static int dh_exportdomparams(void *domparams, OSSL_CALLBACK *param_cb,
-                              void *cbarg)
+static void dh_freedata(void *keydata)
 {
-    DH *dh = domparams;
-    OSSL_PARAM_BLD tmpl;
-    OSSL_PARAM *params = NULL;
-    int ret;
+    DH_free(keydata);
+}
 
-    ossl_param_bld_init(&tmpl);
-    if (dh == NULL
-        || !domparams_to_params(dh, &tmpl)
-        || (params = ossl_param_bld_to_param(&tmpl)) == NULL)
-        return 0;
-    ret = param_cb(params, cbarg);
-    ossl_param_bld_free(params);
-    return ret;
+static int dh_has_domparams(void *keydata)
+{
+    DH *dh = keydata;
+
+    return (DH_get0_p(dh) != NULL && DH_get0_g(dh) != NULL);
+}
+
+static int dh_has_public(void *keydata)
+{
+    DH *dh = keydata;
+
+    return (DH_get0_pub_key(dh) != NULL);
 }
 
-static void *dh_importkey(void *provctx, const OSSL_PARAM params[])
+static int dh_has_private(void *keydata)
 {
-    DH *dh;
-
-    if ((dh = DH_new()) == NULL
-        || !params_to_key(dh, params)) {
-        DH_free(dh);
-        dh = NULL;
-    }
-    return dh;
+    DH *dh = keydata;
+
+    return (DH_get0_priv_key(dh) != NULL);
 }
 
-static int dh_exportkey(void *key, OSSL_CALLBACK *param_cb, void *cbarg)
+static int dh_import(void *keydata, int selection, const OSSL_PARAM params[])
 {
-    DH *dh = key;
+    DH *dh = keydata;
+
+    if (dh == NULL)
+        return 0;
+
+    if (selection != OSSL_KEYMGMT_WANT_KEY
+        && !params_to_domparams(dh, params))
+        return 0;
+    if (selection != OSSL_KEYMGMT_WANT_DOMPARAMS
+        && !params_to_key(dh, params))
+        return 0;
+
+    return 1;
+}
+
+static int dh_export(void *keydata, int selection, OSSL_CALLBACK *param_cb,
+                     void *cbarg)
+{
+    DH *dh = keydata;
     OSSL_PARAM_BLD tmpl;
     OSSL_PARAM *params = NULL;
     int ret;
 
+    if (dh == NULL)
+        return 0;
+
     ossl_param_bld_init(&tmpl);
-    if (dh == NULL
-        || !key_to_params(dh, &tmpl)
-        || (params = ossl_param_bld_to_param(&tmpl)) == NULL)
+
+    if (selection != OSSL_KEYMGMT_WANT_KEY
```

Comment:
OSSL_KEYMGMT_WANT_BOTH ??

Reply:
So you'd prefer the more explicit `if (selection == OSSL_KEYMGMT_WANT_ATTRIBUTES || selection == OSSL_KEYMGMT_WANT_BOTH)`?

Reply:
I would prefer flags so you dont need to do that.

Reply:
`if ((selection & OSSL_KEYMGMT_WANT_PARAMS) != 0)`

Reply:
Yup, that's what I ended up implementing.